### ISSUE TITLE
Support for dtypes other than float64

### ISSUE THREAD

**Issue by c-thiel** [scikit-learn-contrib/hdbscan#108]
Hi,

Version: 0.8.10

I just noted that the following snippet leads to an error as only 64 bit floats are supported:
```python
import hdbscan
import numpy as np
from sklearn.datasets import make_blobs
from sklearn.metrics.pairwise import pairwise_distances
blobs, labels = make_blobs(n_samples=2000, n_features=10)
distance_matrix = pairwise_distances(blobs)
hdb = hdbscan.HDBSCAN(metric='precomputed').fit(distance_matrix.astype(np.float32))
hdb.labels_
```

This is especially a problem for some external metrics. Obviously a typecast to np.float64 fixes the problem, however also doubles the memory usage (which in my case just makes the difference of fitting into RAM)

A quick fix would be to enforce `np.float64` as dtype if `precomputed` metric is selected. However the simple usage of `cython.floating` (or maybe even `cython.numeric`) instead of `np.double_t` should be sufficient to add compatibility on C-level for both dtypes. 

Thanks!

**Comment by lmcinnes** [comment]
Thanks for the suggestion. I'll see what I can do to get that fixed up.

**Comment by lodemo** [comment]
@c-thiel Did you presently found a solution to support float32?

Currently have the same problem with too big precomputed distances matrices.

**Comment by lmcinnes** [comment]
Not at this time, I've been unfortunately busy with many other things and haven't had time to sit down and work through this. I would gladly accept pull requests.

**Comment by c-thiel** [comment]
Yes, I got it to work. Basically you have to interchange the runtime dtype `np.double_t` in the `.pyx` Files with the cython built-in fused type `cython.floating` which supports 32 bit floats and 64 bit doubles and recompile everything. Make sure to leave the `np.double_t` in array initialisations, but change all the static function input types. If you want, you can change the initialisations to `np.float_t` to save even more space due to smaller temporal variables. 

I am happy to prepare a PR, however I am currently on the move and won't be able to do this before October.

**Comment by lmcinnes** [comment]
Thanks @c-thiel that sounds great! I would greatly appreciate it if you would make PR, obviously whenever you get time. I really appreciate the work.
